staging/src/k8s.io/apiserver/pkg/server/filters/reqmgmt.go

```diff
@@ -0,0 +1,264 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"fmt"
+	"net/http"
+	"sync/atomic"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+
+	// TODO: decide whether to use the existing metrics, which
+	// categorize according to mutating vs readonly, or make new
+	// metrics because this filter does not pay attention to that
+	// distinction
+
+	// "k8s.io/apiserver/pkg/endpoints/metrics"
+
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	restclient "k8s.io/client-go/rest"
+	cache "k8s.io/client-go/tools/cache"
+	workqueue "k8s.io/client-go/util/workqueue"
+
+	rmtypesv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+
+	"k8s.io/klog"
+)
+
+// This request filter implements https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+
+// FairQueuingFactory knows how to make FairQueueingSystem objects.
+// This filter makes a FairQueuingSystem for each priority level.
+type FairQueuingFactory interface {
+	NewFairQueuingSystem(concurrencyLimit, numQueues, queueLengthLimit int, requestWaitLimit time.Duration, clk clock.Clock) FairQueuingSystem
+}
+
+// FairQueuingSystem is the abstraction for the queuing and
+// dispatching functionality of one non-exempt priority level.  It
+// covers the functionality described in the "Assignment to a Queue",
+// "Queuing", and "Dispatching" sections of
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190228-priority-and-fairness.md
+// .  Some day we may have connections between priority levels, but
+// today is not that day.
+type FairQueuingSystem interface {
+
+	// The following methods modify or query the configuration
+
+	SetConcurrencyLimit(int)
+	SetDesiredNumQueues(int)
+	GetDesiredNumQueues() int
+	GetActualNumQueues() int
+	SetQueueLengthLimit(int)
+	SetRequestWaitLimit(time.Duration)
+
+	// DoOnEmpty records a function to be called when all the queues
+	// have no requests waiting nor executing.  The filter uses this
+	// for a priority level that has become undesired, setting a
+	// callback that will actually remove the priority level by
+	// establishing a new RMState whose priorityLevelStates map lacks
+	// that entry.  If the priority level becomes desired again before
+	// this is called, the filter cancels by calling `DoOnEmpty(nil)`.
+	// The callback is invoked with no mutexes locked.
+	DoOnEmpty(func())
```

Comment:
>  If the priority level becomes desired again before
>  this is called, the filter cancels by calling `DoOnEmpty(nil)`.

if a PL gets removed and created again instantly, it should be a fresh new priorityLevelStates instance. twiddling the PL between desired and undesired will make the system complex, isn't it? how do you think about rename this to `Destroy(callback func())` to clarify the usage?